File: Python/day07/AOC.py
import re
import time
from functools import lru_cache

def parse_input(input):
    lines = input.split("\n")
    equations = []
    for line in lines:
        nums = list(map(int, re.findall(r"\d+", line)))
        equations.append((nums[0], tuple(nums[1:])))
    return equations

# cache and exiting early if the current sum is greater than the target, reduces the runtime from 7 seconds to 4 seconds
@lru_cache
def evaluate(nums: list[int], target: int, curr=0, index=0, part2=False):
    if curr > target:
        return False
    if index == len(nums):
       return curr == target
   
    num = nums[index]
    
    add = evaluate(nums, target, curr + num, index + 1, part2)
    if add:
        return True
    
    mult = evaluate(nums, target, curr * num if curr != 0 else num, index + 1, part2)
    if mult:
        return True
    
    if part2:
        concat = evaluate(nums, target, int(f'{curr}{num}'), index + 1, part2)
        if concat:
            return True
    
    return False
def part1(input):
    return sum(eq[0] for eq in parse_input(input) if evaluate(eq[1], eq[0]))

def part2(input):
    return sum(eq[0] for eq in parse_input(input) if evaluate(eq[1], eq[0], part2=True))

def main():
    input = open("input.txt", "r").read()
    t1 = time.perf_counter()
    print(f"Part 1: {part1(input)}")
    print(f"Took: {time.perf_counter() - t1:.3f} seconds")
    t1 = time.perf_counter()
    print(f"Part 2: {part2(input)}")
    print(f"Took: {time.perf_counter() - t1:.3f} seconds")

if __name__ == "__main__":
    main()

File: Go/day07/AOC.go
package main

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
	"utils"
)

func parseInput(input []string) [][]int {
	res := make([][]int, len(input))
	for i, line := range input {
		vals := regexp.MustCompile(`\d+`).FindAllString(line, -1)
		nums := make([]int, len(vals))
		for n, num := range vals {
			x, _ := strconv.Atoi(num)
			nums[n] = x
		}
		res[i] = nums
	}
	return res
}

// Brute force all possible combinations of operations using bit manipulation
func part1(input []string) int {
	equations := parseInput(input)
	sum := 0
	for _, eq := range equations {
		target := eq[0]
		eq = eq[1:]
		for n := range int(math.Pow(2, float64(len(eq)-1))) {
			res := eq[0]
			for i := 0; i < len(eq)-1; i++ {
				b := eq[i+1]
				op := byte((n >> i) & 0x1)
				switch op {
				case 0x0:
					res += b
				case 0x1:
					res *= b
				}
				if res > target {
					break
				}
			}
			if res == target {
				// fmt.Printf("%d %d\n", res, target)
				sum += target
				break
			}
		}
	}
	return sum
}

// Brute force all possible combinations of operations using trinary
// runs in about 2.9 seconds
func part2(input []string) int {
	equations := parseInput(input)
	sum := 0
	for _, eq := range equations {
		target := eq[0]
		eq = eq[1:]
		for n := range int(math.Pow(3, float64(len(eq)-1))) {
			res := eq[0]
			for i := 0; i < len(eq)-1; i++ {
				b := eq[i+1]
				op := byte((n / int(math.Pow(3, float64(i)))) % 3)
				switch op {
				case 0x0:
					res += b
				case 0x1:
					res *= b
				case 0x2:
					digits := int(math.Log10(float64(b))) + 1
					concat := int(math.Pow(10, float64(digits)))*res + b
					res = concat
				}
				if res > target { // this saved about 0.4 seconds
					break
				}
			}
			if res == target {
				sum += target
				break
			}
		}
	}
	return sum
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := strings.Split(string(input), "\n")
	utils.Timer(func() {
		fmt.Printf("Part 1: %d\n", part1(lines))
	}, "Part 1")
	utils.Timer(func() {
		fmt.Printf("Part 2: %d\n", part2(lines))
	}, "Part 2")
}
